stdlib/public/core/Algorithm.swift

```diff
@@ -105,9 +105,9 @@ public struct EnumeratedIterator<
   ///
   /// Once `nil` has been returned, all subsequent calls return `nil`.
   public mutating func next() -> Element? {
-    guard let b = _base.next() else { return nil }
-    defer { _count += 1 }
-    return (offset: _count, element: b)
+    let result = _base.next().map { (offset: _count, element: $0) } 
```

Comment:
You can even increment the `_count` inside the closure:

```swift
  return _base.next().map { x in
    let count = _count
    _count += 1
    return (offset: count, element: x)
  }
```

This way instead of 2 branches (one explicit in the next line, and one implicit in the `Optional.map`, there will be only one. Having said that, it's not really a good idea to have a side effect within a closure passed into `map`.

Calling `map` might also have some performance consequences. @swiftix or @gottesmm what do you think?
`defer` will be inlined one way or the other, I don't know about `map`.

All that being said, I really don't find the old code ugly or hard to understand... If only maybe:

```swift
guard let b = _base.next() else { return nil }
let count = _count
_count += 1
return (offset: count, element: b)
```

Which is just a manual de-sugaring of the `defer` block.